### PR TITLE
Drop py27 and py34 support

### PR DESCRIPTION
**Pull request by nicoddemus**
Related to #5275

### CODE REVIEW

**Comment by nicoddemus** [comment]
Oh coverage will drop quite a bit because of dead Python 2 code (which we should tackle in future PRs).

File: tox.ini [hunk 1]
```diff
@@ -56,15 +53,6 @@ deps =
     {env:_PYTEST_TOX_EXTRA_DEP:}
 platform = {env:_PYTEST_TOX_PLATFORM:.*}
 
-[testenv:py27-subprocess]
-deps =
-    pytest-xdist>=1.13
-    py27: mock
-    nose
-commands =
-    pytest -n auto --runpytest=subprocess {posargs}
```
Comment (blueyed) [inline tox.ini]: Should this move into py37-subprocess then?
(But I'd be OK with removing it)

**Comment by blueyed** [comment]
> which we should tackle in future PRs

Why not here?
We have the coverage indicator here, but not later anymore.

**Comment by nicoddemus** [comment]
> Why not here?
We have the coverage indicator here, but not later anymore.

Fair enough. I will get to this tomorrow or after 4.6 gets out. 👍

File: tox.ini (hunk 1 above)
Comment (nicoddemus) [reply tox.ini]: I'm fine removing it too.

File: setup.cfg [hunk 2]
```diff
@@ -23,13 +23,10 @@ classifiers =
...
     Topic :: Software Development :: Libraries
     Topic :: Utilities
-    Programming Language :: Python :: 2
-    Programming Language :: Python :: 2.7
-    Programming Language :: Python :: 3
-    Programming Language :: Python :: 3.4
     Programming Language :: Python :: 3.5
     Programming Language :: Python :: 3.6
     Programming Language :: Python :: 3.7
+    Programming Language :: Python :: 3.8
```
Comment (asottile) [inline setup.cfg]: there's a `Python 3 only` classifier: https://github.com/asottile/flake8-typing-imports/blob/155351518b7786fa0ede18fc48288a1cbf2cc928/setup.cfg#L15

File: setup.py [hunk 3]
```diff
@@ -7,10 +7,8 @@
     "six>=1.10.0",
     "packaging",
     "attrs>=17.4.0",
-    'more-itertools>=4.0.0,<6.0.0;python_version<="2.7"',
```
Comment (RonnyPfannschmidt) [inline setup.py]: why is the max constraint removed?

File: testing/python/integration.py [hunk 4]
```diff
@@ -163,105 +163,6 @@ def test_hello(inject_me):
         reprec = testdir.inline_run()
         reprec.assertoutcome(passed=1)
 
-    def test_unittest_mock_and_pypi_mock(self, testdir):
```
Comment (RonnyPfannschmidt) [inline testing/python/integration.py]: that one seems to help track a misbehavior that users might still see

File: setup.cfg [hunk 5]
```diff
@@ -43,7 +40,7 @@ packages =
     _pytest.mark
 
 py_modules = pytest
-python_requires = >=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*
+python_requires = >=3.5
```
Comment (RonnyPfannschmidt) [inline setup.cfg]: :smile:

File: setup.py (hunk 3 above)
Comment (nicoddemus) [reply setup.py]: It still exists in the line below

File: testing/python/integration.py (hunk 4 above)
Comment (nicoddemus) [reply testing/python/integration.py]: Good point, will bring it back then

**Comment by blueyed** [comment]
Pushed a minor commit removing one instance/import of six - it should get removed completely (added a TODO).

**Comment by asottile** [comment]
pyupgrade will handle that @blueyed -- no need to make those changes here

**Comment by blueyed** [comment]
@asottile 
Ah, sure - should be run here then already, no?

**Comment by asottile** [comment]
it was my understanding from https://github.com/pytest-dev/pytest/issues/5275 that we were going to do that in two separate branches (this one, and then I'd follow up with a separate pyupgrade branch)

**Comment by blueyed** [comment]
It would help / is required for the coverage - otherwise you would have to compare this "manually" from before this PR and after the pyupgrade one.
I think it is good to have this all in one place.

**Comment by asottile** [comment]
:man_shrugging:  I don't think the coverage number matters too much -- I'd prefer two separate PRs personally

**Comment by nicoddemus** [comment]
About the two separate PRs:

One problem with just pushing manual + automatic changes in the same PR is that it makes it difficult to *review* the manual changes. I'm sure if we had just pushed the changes we have here in addition to automatic changes introduced by `pyupgrade`, the comments made here might not have been made at all among all the noise. On the other hand, I understand @blueyed's sentiment about having a clear place to check for the absolute coverage.

Having said that, how about:

1. We keep the scope of the PR as it is now, only the manual changes I've made + removing the `if six.P*` blocks. I plan to finish this today. 
1. Once this is done, people should request further changes and finally approve it.
1. Once everyone has approved this, instead of merging it, @asottile then pushes the automatic changes to this PR. 
1. Everybody now has a chance of requesting further changes until we finally merge it (after 4.6 is out of course).

This might seem complicated at first sight, but is in essence just "approve it now what you see, later we will push more changes before merging". 

What do you guys think?

**Comment by asottile** [comment]
> We keep the scope of the PR as it is now, only the manual changes I've made + removing the `if six.P*` blocks. I plan to finish this today.

pyupgrade is about to support this as well, mostly done :) https://github.com/asottile/pyupgrade/pull/153

**Comment by nicoddemus** [comment]
> pyupgrade is about to support this as well, mostly done :) asottile/pyupgrade#153

Cool!

I will do it myself later anyway, a few of places are not straightforward to remove cleanly. 

Some examples:

https://github.com/pytest-dev/pytest/blob/81cc73103a215c91f0befa4d155f68d08e576140/src/_pytest/pathlib.py#L87-L95

Here it makes sense to use `max` directly instead of `_max`, assuming `pyupgrade` will keep just `_max = max`.

https://github.com/pytest-dev/pytest/blob/81cc73103a215c91f0befa4d155f68d08e576140/src/_pytest/capture.py#L842-L850

Here it makes sense to also remove `_attempt_to_close_capture_file` and call `f.close()` directly instead.

Also there are tests which are Python 2 only which should be removed completely.

There other exmaples, although most would be handled by `pyupgrade` just fine. 😁

File: testing/python/integration.py (hunk 4 above)
Comment (nicoddemus) [reply testing/python/integration.py]: Brought them back, and added `mock` as a testing dependency because of this integration tests.

**Comment by nicoddemus** [comment]
OK, removed everything I could find that was related to Python 2 in the code: code blocks with `sys.version_info`, `six.PY2`, `six.PY3`, `_pytest.compat` functions no longer necessary, etc. 😅

File: src/_pytest/_code/code.py [hunk 6]
```diff
@@ -784,61 +775,52 @@ def _truncate_recursive_traceback(self, traceback):
...
+                descr = "During handling of the above exception, another exception occurred:"
+            else:
+                e = None
+        repr_chain.reverse()
+        return ExceptionChainRepr(repr_chain)
 
 
 class TerminalRepr(object):
     def __str__(self):
         s = self.__unicode__()
```
Comment (asottile) [inline src/_pytest/_code/code.py]: v minor, we should be able to get rid of `__unicode__` for `__str__`

**Comment by nicoddemus** [comment]
> v minor, we should be able to get rid of __unicode__ for __str__

Good catch, missed it. I will get to this later.

**Comment by nicoddemus** [comment]
> I'm going to wait to pyupgrade this until we release 4.6 so we can rebase on top of that

Yes, I was going to suggest that. We should do that right before merging this PR, after 4.6 is out. 👍

File: src/_pytest/compat.py [hunk 7]
```diff
@@ -195,72 +166,50 @@ def _translate_non_printable(s):
...
 
 
-if _PY3:
-    STRING_TYPES = bytes, str
-    UNICODE_TYPES = six.text_type
+STRING_TYPES = bytes, str
+UNICODE_TYPES = six.text_type
 
-    if PY35:
+if PY35:
```
Comment (nicoddemus) [inline src/_pytest/compat.py]: This should be fixed, PY35 is `sys.version_info[:2] >= (3, 5)` so this is always `True`

File: src/_pytest/compat.py [hunk 8]
```diff
@@ -195,72 +166,50 @@ def _translate_non_printable(s):
     return s.translate(_non_printable_ascii_translate_table)
 
 
-if _PY3:
-    STRING_TYPES = bytes, str
-    UNICODE_TYPES = six.text_type
+STRING_TYPES = bytes, str
```
Comment (nicoddemus) [inline src/_pytest/compat.py]: Review where those are used

File: .travis.yml [hunk 9]
```diff
@@ -25,7 +25,7 @@ jobs:
       #       Log: https://travis-ci.org/pytest-dev/pytest/jobs/500358864
       osx_image: xcode10.1
       language: generic
-      env: TOXENV=py37-xdist
+      env: TOXENV=py37-xdist PYTEST_COVERAGE=1
```
Comment (nicoddemus) [inline .travis.yml]: Good catch! 😅

**Comment by nicoddemus** [comment]
Rebased/squased into `master`: we have already created the `4.6-maintenance` branch, so this can go directly to `master`.

Also I think we can run `pyupgrade` in a separate PR, as it should not affect coverage at all.

File: doc/en/example/markers.rst [hunk 10]
```diff
@@ -288,8 +288,7 @@ its test methods:
 This is equivalent to directly applying the decorator to the
 two test functions.
 
-To remain backward-compatible with Python 2.4 you can also set a
-``pytestmark`` attribute on a TestClass like this:
+Due to legacy reasons, setting ``pytestmark`` attribute on a TestClass like this:
```
[the remark this answers is not in the document]
Comment (nicoddemus) [reply doc/en/example/markers.rst]: Indeed, thanks! Done.

**Comment by asottile** [comment]
sweet, let's get this into master :D

File: src/_pytest/compat.py [hunk 11]
```diff
@@ -357,36 +290,6 @@ def safe_isclass(obj):
...
-    From https://docs.python.org/3/library/unittest.html?highlight=unittest#unittest.TestResult.wasSuccessful:
-        Changed in version 3.4: Returns False if there were any
-        unexpectedSuccesses from tests marked with the expectedFailure() decorator.
-    """
-    return sys.version_info >= (3, 4)
-
-
-if _PY3:
-
-    def safe_str(v):
```
Comment (RonnyPfannschmidt) [inline src/_pytest/compat.py]: i noted this one is evil, looking at usage sites made me think the removal was bad in some case

Comment (nicoddemus) [reply src/_pytest/compat.py]: What do you mean @RonnyPfannschmidt? The Python 3 version was just calling `str()`, so I changed the call sites to call `str()` as well.

Comment (RonnyPfannschmidt) [reply src/_pytest/compat.py]: @nicoddemus the function name was terrible for indicating the actual intent

Comment (nicoddemus) [reply src/_pytest/compat.py]: Oh OK.

Could you please "accept" the PR once you finished the review? It would be nice to get this into `master` soon.

**Comment by nicoddemus** [comment]
🎉

**Pull request merged**